### PR TITLE
Add v1 typescript sdk

### PR DESCRIPTION
**Pull request by jarry-xiao**
image

### CODE REVIEW

File: typescript/phoenix-sdk/src/market.ts [hunk 1]
```diff
@@ -0,0 +1,317 @@
...
+  bids: Array<[number, number]>;
+  asks: Array<[number, number]>;
+
+  constructor(asks: Array<[number, number]>, bids: Array<[number, number]>) {
+    this.asks = asks;
+    this.bids = bids;
+  }
+}
+
+export const deserializeMarket = (data: Buffer): Market => {
```
Comment (eugene-chen) [inline typescript/phoenix-sdk/src/market.ts:185]: add comment linking to the struct definition in phoenix-types

File: typescript/phoenix-sdk/src/market.ts [hunk 2]
```diff
@@ -0,0 +1,317 @@
...
+  let bidBuffer = remaining.subarray(offset, offset + bidsSize);
+  offset += bidsSize;
+  let askBuffer = remaining.subarray(offset, offset + asksSize);
+  offset += asksSize;
+  let traderBuffer = remaining.subarray(offset, offset + tradersSize);
+
+  let bidsUnsorted = deserializeTree(bidBuffer, orderIdBeet, restingOrderBeet);
+  let asksUnsorted = deserializeTree(askBuffer, orderIdBeet, restingOrderBeet);
+
+  let bids = [...bidsUnsorted].sort(
```
Comment (eugene-chen) [inline typescript/phoenix-sdk/src/market.ts:226]: does this respect time-ordering for orders at the same price level?

File: typescript/phoenix-sdk/src/market.ts [hunk 3]
```diff
@@ -0,0 +1,317 @@
...
+    takerFeeBps,
+    collectedAdjustedQuoteLotFees,
+    unclaimedAdjustedQuoteLotFees,
+    bids,
+    asks,
+    traders
+  );
+};
+
+function deserializeTree<Key, Value>(
```
Comment (eugene-chen) [inline typescript/phoenix-sdk/src/market.ts:257]: `deserializeRBTree`, also link the sokoban library here

File: typescript/phoenix-sdk/src/market.ts [hunk 4]
```diff
@@ -0,0 +1,317 @@
...
+      buffer.subarray(offset, offset + valueSize)
+    );
+    offset += valueSize;
+    nodes.push([key, value]);
+    freeListPointers.push([index, registers[0]]);
+  }
+
+  let freeNodes = new Set<number>();
+  let indexToRemove = freeListHead - 1;
+  // If there's an infinite loop here, that means that the state is corrupted
```
Comment (eugene-chen) [inline typescript/phoenix-sdk/src/market.ts:302]: should we add a counter? and if we go through this like 10k times then throw?

**Review by eugene-chen — COMMENTED** [review]
Really good stuff.

Let's make sure we add some barebones sample code for how to use this.

File: typescript/phoenix-sdk/src/market.ts [hunk 5]
```diff
@@ -0,0 +1,317 @@
...
+    this.unclaimedAdjustedQuoteLotFees = unclaimedAdjustedQuoteLotFees;
+    this.bids = bids;
+    this.asks = asks;
+    this.traders = traders;
+  }
+
+  getLadder(levels: number): Ladder {
+    let bids: Array<[number, number]> = [];
+    let asks: Array<[number, number]> = [];
+    const quoteLotsPerBaseUnit = 10 ** toNum(this.header.quoteParams.decimals);
```
[the remark this answers is not in the document]
Comment (jarry-xiao) [reply typescript/phoenix-sdk/src/market.ts:107]: should be quotAtomsPerBaseUnit, good call

File: typescript/phoenix-sdk/src/market.ts (hunk 2 above)
Comment (jarry-xiao) [reply typescript/phoenix-sdk/src/market.ts:226]: I think it should

Comment (jarry-xiao) [reply typescript/phoenix-sdk/src/market.ts:226]: Jk it doesn't, don't think it's that important for now, will add a TODO

File: typescript/phoenix-sdk/src/market.ts [hunk 6]
```diff
@@ -0,0 +1,341 @@
...
+    this.unclaimedAdjustedQuoteLotFees = unclaimedAdjustedQuoteLotFees;
+    this.bids = bids;
+    this.asks = asks;
+    this.traders = traders;
+  }
+
+  getLadder(levels: number): Ladder {
+    let bids: Array<[number, number]> = [];
+    let asks: Array<[number, number]> = [];
+    const quoteAtomsPerBaseUnit = 10 ** toNum(this.header.quoteParams.decimals);
```
Comment (eugene-chen) [inline typescript/phoenix-sdk/src/market.ts:107]: isn't this `quoteAtomsPerQuoteUnit`?

**Review by eugene-chen — APPROVED** [review]

**Pull request merged**